easybuild/easyblocks/x/xmipp.py

```diff
@@ -25,174 +25,203 @@
 """
 EasyBuild support for building and installing Xmipp, implemented as an easyblock
 
-@author: Jens Timmerman (Ghent University)
-@author: Pablo Escobar (sciCORE, SIB, University of Basel)
-@author: Kenneth Hoste (Ghent University)
+@author: Ake Sandgren (HPC2N, Umea University)
 """
-import fileinput
 import os
-import re
-import stat
-import sys
 
 import easybuild.tools.environment as env
-import easybuild.tools.toolchain as toolchain
-from easybuild.easyblocks.generic.pythonpackage import det_pylibdir
-from easybuild.framework.easyblock import EasyBlock
-from easybuild.framework.easyconfig import CUSTOM
+from distutils.version import LooseVersion
+from easybuild.easyblocks.generic.scons import SCons
 from easybuild.tools.build_log import EasyBuildError
-from easybuild.tools.filetools import adjust_permissions, mkdir, write_file
+from easybuild.tools.filetools import apply_regex_substitutions, change_dir
+from easybuild.tools.filetools import copy_file, mkdir, remove_file, symlink
 from easybuild.tools.modules import get_software_root, get_software_version
 from easybuild.tools.run import run_cmd
+from easybuild.tools.systemtools import get_shared_lib_ext
 
 
-class EB_Xmipp(EasyBlock):
-    """
-    easyblock to install Xmipp
-    """
+class EB_Xmipp(SCons):
+    """Support for building/installing Xmipp."""
 
     def __init__(self, *args, **kwargs):
-        """Easyblock constructor, enable building in installation directory."""
+        """Initialize Xmipp-specific variables."""
         super(EB_Xmipp, self).__init__(*args, **kwargs)
-        self.build_in_installdir = True
-        self.xmipp_pythonpaths = []
+        self.xmipp_modules = ['xmippCore', 'xmipp', 'xmippViz']
+        self.srcdir = os.path.join(self.builddir, 'src')
+        self.cfg['start_dir'] = self.builddir
+        self.use_cuda = False
 
     def extract_step(self):
         """Extract Xmipp sources."""
-        # strip off 'xmipp' part to avoid having everything in a 'xmipp' subdirectory
-        self.cfg.update('unpack_options', '--strip-components=1')
+        # Xmipp assumes that everything is unpacked in a "src" dir
+        mkdir(self.srcdir)
+        self.cfg.update('unpack_options', '--directory src')
         super(EB_Xmipp, self).extract_step()
+        for module in self.xmipp_modules:
+            symlink('%s-%s' % (module, self.version), os.path.join(self.srcdir, module), use_abspath_source=False)
+
+    def patch_step(self):
+        """Patch files from self.srcdir dir."""
+        super(EB_Xmipp, self).patch_step(beginpath=self.srcdir)
+
+    def setup_xmipp_env(self):
+        """Setup environment before running SCons."""
+        env.setvar('LINKERFORPROGRAMS', os.environ['CXX'])
+        env.setvar('MPI_CC', os.getenv('MPICC'))
+        env.setvar('MPI_CXX', os.getenv('MPICXX'))
+        env.setvar('MPI_LINKERFORPROGRAMS', os.getenv('MPICXX'))
+        incpath = os.path.join(get_software_root('Java'), 'include')
```

Comment:
This will crash hard if `Java` is not included as a dep, should be handled with:

```python
java_root = get_software_root('Java')
if java_root:
    java_incpath = os.path.join(java_root, 'include')
else:
    raise EasyBuildError("Java not included as dependency")
```

Reply:
It won't since that function is only used after veryfying that the required deps (of which Java is one) are already there.

See the deps struct and the missing_deps loop.

Reply:
OK, I failed to take into account the order of methods being called